File: app/src/main/java/com/blogspot/spartandeveloper/playlistmessagesforspotify/data/PlaylistCreator.java
package com.blogspot.spartandeveloper.playlistmessagesforspotify.data;

import com.blogspot.spartandeveloper.playlistmessagesforspotify.util.events.CreatePlaylistErrorEvent;
import com.blogspot.spartandeveloper.playlistmessagesforspotify.util.events.CreatePlaylistSuccessEvent;
import com.blogspot.spartandeveloper.playlistmessagesforspotify.util.events.LoadPlaylistsEvent;

import org.greenrobot.eventbus.EventBus;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import kaaes.spotify.webapi.android.SpotifyService;
import kaaes.spotify.webapi.android.models.Pager;
import kaaes.spotify.webapi.android.models.Playlist;
import kaaes.spotify.webapi.android.models.PlaylistTrack;
import kaaes.spotify.webapi.android.models.Track;
import kaaes.spotify.webapi.android.models.TracksPager;
import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;
import timber.log.Timber;

class PlaylistCreator {

    private String userId, playlistName, message;
    private SpotifyService spotify;
    private int totalTracks, runningTrackCount = 0;
    private ArrayList<Track> tracks;

    PlaylistCreator(String userId, SpotifyService spotify, String playlistName, String message) {
        this.userId = userId;
        this.spotify = spotify;
        this.playlistName = playlistName;
        this.message = message;
    }

    void execute() {

        String[] split = message.split(" ");
        totalTracks = split.length;
        tracks = new ArrayList<>(totalTracks);

        for (int i = 0; i < totalTracks; i++) {
            tracks.add(null);
        }

        // loop over all tracks
        for (int i = 0; i < totalTracks; i++) {
            String curr = split[i];
            searchTrack(curr, i, 0);
        }

    }


    private void searchTrack(final String curr, final int index, final int offset) {
        Map<String, Object> options = new HashMap<>();
        options.put("limit", 50);
        options.put("offset", offset);
        Timber.d("searchTracks. curr:%s, offset:%s, index:%s", curr, offset, index);

        spotify.searchTracks(curr, options, new Callback<TracksPager>() {
            @Override
            public void success(TracksPager tracksPager, Response response) {

                for (Track track : tracksPager.tracks.items) {
                    String trackName = track.name;
                    if (trackName.equalsIgnoreCase(curr)) {
                        Timber.i("found word: %s", trackName);
                        tracks.set(index, track);
                        runningTrackCount++;
                        checkAndHandleCompletion();
                        return;
                    }
                }

                if (offset <= 200) {
                    // search next 50
                    searchTrack(curr, index, offset + 50);
                } else {
                    // could not find string
                    EventBus.getDefault().post(new CreatePlaylistErrorEvent(curr));
                }

            }

            @Override
            public void failure(RetrofitError error) {
                Timber.e("retrofit error finding track");
                Timber.e(error);
            }
        });


    }

    private void checkAndHandleCompletion() {
        Timber.d("checkAndHandleCompletion. runningTrackCount:%s, totalTracks:%s", runningTrackCount, totalTracks);
        if (runningTrackCount < totalTracks) return;

        Map<String, Object> options = new HashMap<>();
        options.put("name", playlistName);
        spotify.createPlaylist(userId, options, new Callback<Playlist>() {
            @Override
            public void success(final Playlist playlist, Response response) {
                Timber.i("created playlist: %s", playlist.name);

                final Map<String, Object> addTrackBody = new HashMap<>();
                String[] uris = new String[tracks.size()];
                for (int i = 0; i < tracks.size(); i++) {
                    uris[i] = tracks.get(i).uri;
                }
                addTrackBody.put("uris", uris);

                spotify.addTracksToPlaylist(userId, playlist.id, null, addTrackBody,
                        new Callback<Pager<PlaylistTrack>>() {
                            @Override
                            public void success(Pager<PlaylistTrack> playlistTrackPager, Response response) {
                                Timber.d("added tracks to playlist: %s", playlist.name);
                                Timber.d("response: %s", response.getReason());
                                EventBus.getDefault().post(new LoadPlaylistsEvent());
                                EventBus.getDefault().post(new CreatePlaylistSuccessEvent(playlistName));
                            }

                            @Override
                            public void failure(RetrofitError error) {
                                Timber.e("unable to add tracks to playlist");
                                Timber.e(error);
                            }
                        });
            }

            @Override
            public void failure(RetrofitError error) {
                Timber.e("unable to create playlist");
                Timber.e(error);
            }
        });



    }


}

File: app/src/main/java/com/blogspot/spartandeveloper/playlistmessagesforspotify/data/PlaylistCreator.kt
package com.blogspot.spartandeveloper.playlistmessagesforspotify.data

import com.blogspot.spartandeveloper.playlistmessagesforspotify.util.events.CreatePlaylistErrorEvent
import com.blogspot.spartandeveloper.playlistmessagesforspotify.util.events.CreatePlaylistSuccessEvent
import com.blogspot.spartandeveloper.playlistmessagesforspotify.util.events.LoadPlaylistsEvent
import kaaes.spotify.webapi.android.SpotifyService
import kaaes.spotify.webapi.android.models.Track
import org.greenrobot.eventbus.EventBus
import timber.log.Timber
import java.util.*

internal class PlaylistCreator (val spotify: SpotifyService, val playlistName: String, val message: String) {

    private val MAX_SPOTIFY_OFFSET = 450
    private val INC_SPOTIFY_OFFSET = 50

    fun execute(): Boolean {

        val split = message.trim().split(" ")

        val it = split.iterator()
        val userQuery = LinkedList<String>()
        while (it.hasNext()) userQuery.addLast(it.next())
        val result = LinkedList<Track>()

        if (!executeUtil(userQuery = userQuery, result = result)) {
            // FIXME change err msg
            EventBus.getDefault().post(CreatePlaylistErrorEvent(""))
            return false
        } else {
            EventBus.getDefault().post(LoadPlaylistsEvent())
            EventBus.getDefault().post(CreatePlaylistSuccessEvent(playlistName))
            return true
        }

    }

    private fun executeUtil(userQuery: MutableList<String>, startIndex: Int = 0, result: MutableList<Track>): Boolean {

        // finished; no any remaining words
        if (startIndex == userQuery.size) {
            return true
        }

        for (i in startIndex until userQuery.size) {

            // create query for search range
            val searchList = userQuery.subList(startIndex, i + 1)
            val sb = StringBuilder()
            for (s: String in searchList) {
                sb.append(s).append(" ")
            }
            val subQuery = sb.toString().trim()

            val foundTrack = search(subQuery)

            if (foundTrack != null) {

                // track successes
                result.add(foundTrack)

                // bubble up successful query
                if (executeUtil(userQuery, i + 1 , result)) {
                    return true
                }

                // backtrack for failure
                result.remove(foundTrack)
            }
        }

        return false
    }

    private fun search(query: String): Track? {

        var offset = 0
        val options = HashMap<String, Any>()
        options["limit"] = 50
        options["offset"] = offset

        // paginated search until found query or max offset
        while (offset <= MAX_SPOTIFY_OFFSET) {

            val pager = spotify.searchTracks(query, options)

            for (track in pager.tracks.items) {
                val trackName = track.name
                if (trackName.equals(query, ignoreCase = true)) {
                    Timber.i("found word: %s", trackName)
                    return track
                }
            }

            offset += INC_SPOTIFY_OFFSET
            options["offset"] = offset

        }

        return null
    }

}